### PR TITLE
chat 1 vs 1

### PR DESCRIPTION
**Pull request by ThuyenPVT**

### CODE REVIEW

File: lib/data/entities/message.dart [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+class Message {
+  final String senderID;
+  final String content;
+  final Timestamp timestamp;
+  final MessageType type;
+
+  Message({this.senderID, this.content, this.timestamp, this.type});
+}
```
Comment (chuongdevJC) [inline lib/data/entities/message.dart]: .

File: lib/pages/recent_conversation/conversation_page.dart [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+  Widget _conversationPageUI(
+    BuildContext _context,
+    Conversation conversations,
+  ) {
+    return _buildList(conversations);
+  }
+
+  Widget _buildList(Conversation conversation) {
+    if (conversation != null) {
+      if (conversation.messages.length != 0) {
```
Comment (chuongdevJC) [inline lib/pages/recent_conversation/conversation_page.dart]: extract

File: lib/bloc/blocs/conversation_bloc.dart [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+      yield FailureConversation();
+    }
+  }
+
+  Stream<ConversationState> _mapConversationToState(
+    String conversationID,
+  ) async* {
+    yield LoadingConversation();
+    try {
+      final _getCurrentUser =
```
Comment (truyenjc) [inline lib/bloc/blocs/conversation_bloc.dart]: name param must be a noun

File: lib/bloc/blocs/conversation_bloc.dart [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+      yield SuccessConversation(conversation: _getCurrentUser);
+    } catch (_) {
+      yield FailureConversation();
+    }
+  }
+
+  Stream<ConversationState> _mapSendMessageToState(
+    String senderID,
+    String conversationID,
+    dynamic message,
```
Comment (truyenjc) [inline lib/bloc/blocs/conversation_bloc.dart]: avoid use dynamic for param type

File: lib/bloc/blocs/conversation_bloc.dart [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+      final _message = Message(
+        senderID: senderID,
+        content: message is File ? _imageURL : message,
+        timestamp: timestamp,
+        type: type,
+      );
+
+      await _conversationRepository.sendMessage(conversationID, _message);
+
+      final _getCurrentUser =
```
Comment (truyenjc) [inline lib/bloc/blocs/conversation_bloc.dart]: like above

File: lib/bloc/states/friend_state.dart [hunk 6]
```diff
@@ -1,5 +1,6 @@
 import 'package:equatable/equatable.dart';
 import 'package:structure_flutter/data/entities/account.dart';
+import 'package:structure_flutter/data/entities/friend_request.dart';
```
Comment (truyenjc) [inline lib/bloc/states/friend_state.dart]: unused import

File: lib/data/source/remote/conversation_remote_datasource.dart [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+        .toList();
+  }
+
+  @override
+  Future<Conversation> getConversations(String conversationID) async {
+    try {
+      final _docsSnapshot =
+          await _conversationCollection.doc(conversationID).get();
+      final list = Conversation.fromFireStore(_docsSnapshot);
+      return list;
```
Comment (truyenjc) [inline lib/data/source/remote/conversation_remote_datasource.dart]:
```suggestion
      return Conversation.fromFireStore(_docsSnapshot);
```

File: lib/data/source/remote/storage_remote_datasource.dart [hunk 8]
```diff
@@ -22,4 +24,15 @@ class StorageRemoteDataSourceImpl extends StorageRemoteDataSource {
...
+  @override
+  Future<String> getDownloadURL() async {
+    try {
+      final _firebaseStorage = _baseRef.child(_profileImages);
+      final url = await _firebaseStorage.getDownloadURL();
+      return url;
+    } catch (_) {
+      return null;
+    }
+  }
```
Comment (truyenjc) [inline lib/data/source/remote/storage_remote_datasource.dart]:
```suggestion
  Future<String> getDownloadURL() {
    try {
      final _firebaseStorage = _baseRef.child(_profileImages);
      return _firebaseStorage.getDownloadURL();
    } catch (_) {
      return null;
    }
  }
```

File: lib/pages/recent_conversation/conversation_page.dart [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
```
Comment (truyenjc) [inline lib/pages/recent_conversation/conversation_page.dart]:
```suggestion
      if (conversation.messages.isNotEmpty) {
```

File: lib/pages/recent_conversation/widgets/bottom_icon.dart [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    this.textFieldController,
+    this.onSelectedEmoji,
+    this.onSendButton,
+    this.onTapEmoji,
+  });
+
+  @override
+  Widget build(BuildContext context) {
+    return Container(
+      // decoration: BoxDecoration(color: AppColors.),
```
Comment (truyenjc) [inline lib/pages/recent_conversation/widgets/bottom_icon.dart]: .

File: lib/pages/recent_conversation/recent_conversation.dart [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+              ),
+            ));
+      },
+      leading: CircleAvatar(
+          radius: 26,
+          backgroundColor: AppColors.blueAccent,
+          child: CircleAvatar(
+            radius: 25,
+            backgroundColor: AppColors.double_spanish_white,
+            backgroundImage: NetworkImage('${conversation.image}'),
```
Comment (truyenjc) [inline lib/pages/recent_conversation/recent_conversation.dart]:
```suggestion
            backgroundImage: NetworkImage(conversation.image),
```

File: lib/pages/recent_conversation/conversation_page.dart [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+  void initState() {
+    _conversationBloc.add(InitConversation(
+      conversationID: widget.conversationID,
+    ));
+
+    textFieldController = TextEditingController()
+      ..addListener(() {
+        setState(() {
+          _message = textFieldController.text;
+        });
```
Comment (truyenjc) [inline lib/pages/recent_conversation/conversation_page.dart:84]:
```suggestion
        setState(() => _message = textFieldController.text);
```

**Review by truyenjc — APPROVED** [review]
OK!

**Pull request merged**